### PR TITLE
Added handling of cases with type overflow

### PR DESCRIPTION
**Pull request by RomanReznichenkoLuxoft**
Fixes #[12381](https://adc.luxoft.com/jira/browse/FORDTCN-12381)

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Summary
SDL crashes when the HMI sends UI.OnTouchEvent with the 'ts' param has a value greater than maxvalue.
I've added processing for the case where the sdl gets numbers that go beyond the type.

### CODE REVIEW

**Review by ychernysheva — APPROVED** [review]

File: src/components/smart_objects/include/smart_objects/number_schema_item.h [hunk 1]
```diff
@@ -157,10 +157,17 @@ errors::eType TNumberSchemaItem<NumberType>::validate(
...
+
   NumberType value(0);
   if (typeid(int32_t) == typeid(value)) {
+    if (Object.asInt() > std::numeric_limits<int32_t>::max() ||
+        Object.asInt() < std::numeric_limits<int32_t>::min())
+      return errors::OUT_OF_RANGE;
     value = utils::SafeStaticCast<int64_t, int32_t>(Object.asInt());
   } else if (typeid(uint32_t) == typeid(value)) {
+    if (Object.asInt() > std::numeric_limits<uint32_t>::max() ||
+        Object.asInt() < std::numeric_limits<int32_t>::min())
```
Comment (VadymLuchko) [inline src/components/smart_objects/include/smart_objects/number_schema_item.h:169]: Seems there we need to use `Object.asUInt()` 
and also std::numeric_limits<**u**int32_t>::min()

**Review by VadymLuchko — CHANGES_REQUESTED** [review]

File: src/components/smart_objects/include/smart_objects/number_schema_item.h (hunk 1 above)
Comment (RomanReznichenkoLuxoft) [reply src/components/smart_objects/include/smart_objects/number_schema_item.h]: Done

**Review by VadymLuchko — APPROVED** [review]

File: src/components/smart_objects/include/smart_objects/number_schema_item.h [hunk 2]
```diff
@@ -157,10 +157,17 @@ errors::eType TNumberSchemaItem<NumberType>::validate(
     report->set_validation_info(validation_info);
     return errors::INVALID_VALUE;
   }
+
   NumberType value(0);
   if (typeid(int32_t) == typeid(value)) {
+    if (Object.asInt() > std::numeric_limits<int32_t>::max() ||
+        Object.asInt() < std::numeric_limits<int32_t>::min())
+      return errors::OUT_OF_RANGE;
```
Comment (AKalinich-Luxoft) [inline src/components/smart_objects/include/smart_objects/number_schema_item.h:165]: @RomanReznichenkoLuxoft in [your case](https://google.github.io/styleguide/cppguide.html#Conditionals) you have to add braces for return statement

File: src/components/smart_objects/include/smart_objects/number_schema_item.h [hunk 3]
```diff
@@ -157,10 +157,17 @@ errors::eType TNumberSchemaItem<NumberType>::validate(
...
     value = utils::SafeStaticCast<int64_t, int32_t>(Object.asInt());
   } else if (typeid(uint32_t) == typeid(value)) {
+    if (Object.asUInt() > std::numeric_limits<uint32_t>::max() ||
+        Object.asUInt() < std::numeric_limits<uint32_t>::min())
+      return errors::OUT_OF_RANGE;
```
Comment (AKalinich-Luxoft) [inline src/components/smart_objects/include/smart_objects/number_schema_item.h:170]: @RomanReznichenkoLuxoft same as above

File: src/components/smart_objects/include/smart_objects/number_schema_item.h [hunk 4]
```diff
@@ -157,10 +157,17 @@ errors::eType TNumberSchemaItem<NumberType>::validate(
...
     value = utils::SafeStaticCast<uint64_t, uint32_t>(Object.asUInt());
```
Comment (AKalinich-Luxoft) [inline src/components/smart_objects/include/smart_objects/number_schema_item.h:171]: @RomanReznichenkoLuxoft do we still need this function? Probably we can simply use `static_cast` here?

**Review by AKalinich-Luxoft — APPROVED** [review]

File: tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py [hunk 5]
```diff
@@ -840,7 +840,28 @@ def _gen_schema_item_decl_code(self, param, member_name, default_value):
                     [[u"bool", None if param.default_value is None
                       else u"true" if param.default_value is True else u"false"]]))
         elif type(param) is Integer:
-            if not param.max_value or param.max_value and param.max_value < 2 ** 31:
+            if param.min_value is None and param.max_value is None:
```
Comment (AKalinich-Luxoft) [inline tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py:843]: @RomanReznichenkoLuxoft `int64_t` should be generated not only in this case. It should be generated when:
- Both man and max are null
- Min is null
- Min less than min(int32_t)
- Max is null or >= max(int32_t) and min < 0

File: tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py [hunk 6]
```diff
@@ -840,19 +840,33 @@ def _gen_schema_item_decl_code(self, param, member_name, default_value):
...
+                         [u"uint64_t", param.max_value],
+                         [u"uint64_t", param.default_value]]))
+            elif (param.min_value is not None and param.min_value >= 0) and (param.max_value is not None and param.max_value < 2 ** 32):
+                code = self._impl_code_integer_item_template.substitute(
+                    type=u"uint32_t",
+                    params=self._gen_schema_item_param_values(
+                        [[u"uint32_t", param.min_value],
+                         [u"uint32_t", param.max_value],
+                         [u"uint32_t", param.default_value]]))
+            elif  (param.min_value is None or param.min_value >= -(2 ** 31) and param.min_value < 0) and (param.max_value is not None and param.max_value < 2 ** 31):
```
Comment (AKalinich-Luxoft) [inline tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py:857]:
```suggestion
            elif  (param.min_value is not None and param.min_value >= -(2 ** 31) and param.min_value < 0) and (param.max_value is not None and param.max_value < 2 ** 31):
```

File: tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py [hunk 7]
```diff
@@ -840,19 +840,33 @@ def _gen_schema_item_decl_code(self, param, member_name, default_value):
...
                 code = self._impl_code_integer_item_template.substitute(
                     type=u"int32_t",
                     params=self._gen_schema_item_param_values(
-                        [[u"int32_t", param.min_value],
+                        [[u"int32_t", param.min_value], 
                          [u"int32_t", param.max_value],
                          [u"int32_t", param.default_value]]))
-            elif param.max_value < 2 ** 63:
+            elif (param.min_value is None or param.min_value < -(2 ** 31) ) and (param.max_value is None or param.max_value < 2 ** 63):
```
Comment (AKalinich-Luxoft) [inline tools/InterfaceGenerator/generator/generators/SmartFactoryBase.py:864]:
```suggestion
            elif (param.min_value is None or param.min_value > -(2 ** 63) ) and (param.max_value is None or param.max_value < 2 ** 63):
```

**Review by VladSemenyuk — APPROVED** [review]

**Pull request closed without merging**